GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/UpgradeOrchestratorWithIProductUpgraderTests.cs

```diff
@@ -0,0 +1,202 @@
+﻿using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Tests.Should;
+using GVFS.UnitTests.Mock.Common;
+using GVFS.UnitTests.Windows.Mock.Upgrader;
+using GVFS.Upgrader;
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+
+namespace GVFS.UnitTests.Upgrader
+{
+    [TestFixture]
+    public class UpgradeOrchestratorWithIProductUpgraderTests
+    {
+        private const string OlderThanLocalVersion = "1.0.17000.1";
+        private const string LocalGVFSVersion = "1.0.18115.1";
+        private const string NewerThanLocalVersion = "1.1.18115.1";
+
+        private delegate void TryGetNewerVersionCallback(out System.Version version, out string message);
+        private delegate void UpgradeAllowedCallback(out string message);
+        private delegate void TryRunPreUpgradeChecksCallback(out string delegateMessage);
+        private delegate void TryDownloadNewestVersionCallback(out string message);
+        private delegate void TryRunInstallerCallback(InstallActionWrapper installActionWrapper, out string error);
+
+        private MockTracer Tracer { get; set; }
```

Comment:
One pattern I wanted to see if we could improve - we use the `Mock` prefix to indicate both a `mocked` object, as well as the `Mock<..>` container for the object. I thought it would be nice if we had some way to distinguish these when looking at the code, so we would know whether what type we are dealing with (and, do we need to pass in the `.Object` or just the `Object` itself.

Reply:
Jameson and I discussed about Mock vs Moq. I think Moq clear-cut. Using Moq.